nemo_skills/code_execution/math_grader.py

```diff
@@ -12,535 +12,214 @@
-            if len(pred_matrix) == len(ref_matrix_items):
-                if all(
-                    [
-                        math_equal(ref, pred, include_percentage, tolerance)
-                        for ref, pred in zip(ref_matrix_items, pred_matrix)
-                    ]
-                ):
-                    return True
-        except Exception:
-            pass
-
-    return symbolic_equal(prediction, reference, tolerance, timeout)
-
-
-def symbolic_equal(a, b, tolerance, timeout=10.0):
-    import sympy
-    from sympy.parsing.latex import parse_latex
-    from sympy.parsing.sympy_parser import parse_expr
-
-    def _parse(s):
-        for f in [parse_expr, parse_latex]:
-            try:
-                with time_limit(timeout):
-                    return f(s)
-            except Exception:
-                pass
-        return s
+            predicted_answer = int(predicted_answer) % take_modulo
+        except:
+            predicted_answer = None
+        # no need to simpy call in this case
+        return predicted_answer == gt_answer
+
+    # Try to compare as MCQ options
+    mcq_options = "ABCDEFGHIJ"
+    norm_gt_mcq = gt_answer.strip()
+
+    is_mcq = re.fullmatch("|".join(mcq_options), norm_gt_mcq)
+    if is_mcq:
+        parsed_gt = parse(gt_answer, [StringExtractionConfig(strings=tuple(mcq_options))])
+        parsed_pred = parse(predicted_answer, [StringExtractionConfig(strings=tuple(mcq_options))])
+        return verify(parsed_gt, parsed_pred)
+    
+    # Additional normalization step
+    gt_answer = _additional_normalization(gt_answer)
+    predicted_answer = _additional_normalization(predicted_answer)
+    
+    # Try literal comparison
+    literal_pattern = r"[a-zA-Z ,]+|[0-9 ]+"
+    normalized_gt = normalize_latex(gt_answer, NormalizationConfig)
+    normalized_pred = normalize_latex(predicted_answer, NormalizationConfig)
+    is_literal = (re.fullmatch(literal_pattern, normalized_gt) and
+                  re.fullmatch(literal_pattern, normalized_pred))
+
+    if is_literal:
+        return normalized_gt.replace(" ", "") == normalized_pred.replace(" ", "")
+
+    # Fallback to symbolic comparison
+    current_gt_answer = gt_answer
+    current_predicted_answer = predicted_answer
+    
+    # math_verify.parse expects input to be in latex environment, e.g. $...$
+    latex_env_search_pattern = r"\$.*\$|\\\(.*\\\)|\\\[.*\\\]|\\boxed\{"
+    if not re.search(latex_env_search_pattern, current_gt_answer, re.DOTALL):
+        current_gt_answer = f"${current_gt_answer}$"
+    if not re.search(latex_env_search_pattern, current_predicted_answer, re.DOTALL):
+        current_predicted_answer = f"${current_predicted_answer}$"
+
+    parsed_gt = parse(current_gt_answer, [LatexExtractionConfig()])
+    parsed_pred = parse(current_predicted_answer, [LatexExtractionConfig()])
+
+    return verify(parsed_gt, parsed_pred, **kwargs)
+
+
+def batch_evaluate_results(
```

Comment:
If this can't use threads, then let's maybe just have a basic loop without saving tmp files and map_to_future logic? So just compute all sequentially, store results in memory and add back to original files after all are computed. If this makes the implementation simpler and easier to read, probably worth doing